# Assignment: Build a Regression Model in Keras

Install Requored Libraries 


In [1]:
# Install all Libraries required for this assignment are listed below. 

!pip install numpy==2.0.2
!pip install pandas==2.2.2
!pip install tensorflow==2.18.0
!pip install scikit-learn


[notice] A new release of pip is available: 23.2.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.2.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.2.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.2.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


# A. Build a baseline model (5 marks) 

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the 
train_test_split
helper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

Submit your Jupyter Notebook with your code and comments.


In [2]:
# import the Keras, Pandas, and Numpy Libraries
import pandas as pd
import numpy as np
import keras
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Download the dataset into pandas library 
# Please download the data to your project folder before calling this code 

filepath='concrete_data.csv'
concrete_data = pd.read_csv(filepath)

concrete_data.head()


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
# Read data column names
concrete_data_columns = concrete_data.columns
print(concrete_data_columns)

# Split data into predictors and target data
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column




Index(['Cement', 'Blast Furnace Slag', 'Fly Ash', 'Water', 'Superplasticizer',
       'Coarse Aggregate', 'Fine Aggregate', 'Age', 'Strength'],
      dtype='object')


In [4]:
# Import Keras packages
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Input


In [5]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Input(shape=(8,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [6]:
# build the model
model = regression_model()

# Split the data into training and test sets
# Train the model on the training data using 50 epochs
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)


model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=2)




Epoch 1/50
23/23 - 0s - 15ms/step - loss: 894068.6250 - val_loss: 719990.5625
Epoch 2/50
23/23 - 0s - 2ms/step - loss: 577362.3125 - val_loss: 461975.0312
Epoch 3/50
23/23 - 0s - 2ms/step - loss: 367028.2500 - val_loss: 294562.2500
Epoch 4/50
23/23 - 0s - 2ms/step - loss: 232416.1406 - val_loss: 187588.6094
Epoch 5/50
23/23 - 0s - 2ms/step - loss: 146392.0312 - val_loss: 118227.7812
Epoch 6/50
23/23 - 0s - 2ms/step - loss: 90868.9297 - val_loss: 73260.1328
Epoch 7/50
23/23 - 0s - 2ms/step - loss: 55451.5664 - val_loss: 44732.9141
Epoch 8/50
23/23 - 0s - 2ms/step - loss: 31985.2949 - val_loss: 22767.0449
Epoch 9/50
23/23 - 0s - 2ms/step - loss: 13772.6230 - val_loss: 8885.8408
Epoch 10/50
23/23 - 0s - 2ms/step - loss: 8144.8696 - val_loss: 7748.4829
Epoch 11/50
23/23 - 0s - 2ms/step - loss: 7555.2148 - val_loss: 7415.3696
Epoch 12/50
23/23 - 0s - 2ms/step - loss: 7198.9087 - val_loss: 6982.9824
Epoch 13/50
23/23 - 0s - 2ms/step - loss: 6871.2607 - val_loss: 6661.1162
Epoch 14/50
23/23 -

In [7]:
# List to store mean squared errors
mse_list = []

# Repeat the split and train steps 50 times
for i in range(50):
    # Split the data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    
    # Create and train the model
    model = regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    
    # Predict on the test set
    y_pred = model.predict(X_test)
    # Calculate mean squared error
    mse = mean_squared_error(y_test, y_pred)  # Now this will work
    mse_list.append(mse)

# Print the list of mean squared errors
print(mse_list)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 


In [8]:
#  Report the mean and the standard deviation of the mean squared errors.

mean_mse = np.mean(mse_list)  # Calculate the mean of the MSEs
std_mse = np.std(mse_list)    # Calculate the standard deviation of the MSEs

# Print the results
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

Mean of Mean Squared Errors: 385.07478211603893
Standard Deviation of Mean Squared Errors: 575.4091683716704


# B. Normalize the data (5 marks) 

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [9]:

# normalize the predictors data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

# List to store mean squared errors
mse_list = []


# Repeat the split and train steps 50 times
for i in range(50):
    # Split the data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    
    # Create and train the model
    model = regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    
    # Predict on the test set
    y_pred = model.predict(X_test)
    
    # Calculate mean squared error
    mse = mean_squared_error(y_test, y_pred)  # Now this will work
    mse_list.append(mse)

# Print the list of mean squared errors
print(mse_list)

#  Report the mean and the standard deviation of the mean squared errors.

mean_mse = np.mean(mse_list)  # Calculate the mean of the MSEs
std_mse = np.std(mse_list)    # Calculate the standard deviation of the MSEs

# Print the results
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 


#### How does the mean of the mean squared errors compare to that from Step A?

#### Answer: the mean of the mean squared errors has been reduced from 385.07 to 372.045

# C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.



In [10]:

# Repeat the split and train steps 100 times
mse_list = []

for i in range(50):
    # Split the data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    
    # Create and train the model
    model = regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=0)
    
    # Predict on the test set
    y_pred = model.predict(X_test)
    
    # Calculate mean squared error
    mse = mean_squared_error(y_test, y_pred)  # Now this will work
    mse_list.append(mse)

# Print the list of mean squared errors
print(mse_list)

#  Report the mean and the standard deviation of the mean squared errors.

mean_mse = np.mean(mse_list)  # Calculate the mean of the MSEs
std_mse = np.std(mse_list)    # Calculate the standard deviation of the MSEs

# Print the results
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 


#### How does the mean of the mean squared errors compare to that from Step B?
#### the mean has been signgactly reduced to 167.705

# D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

In [11]:
# increase the number of hiddern layers to 3 each of 10 nodes and ReLU activation function.
mse_list = []

def regression_model():
    # create model
    model = Sequential()
    model.add(Input(shape=(8,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model
# Repeat the split and train steps 50 times
for i in range(50):
    # Split the data into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    
    # Create and train the model
    model = regression_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    
    # Predict on the test set
    y_pred = model.predict(X_test, verbose=0)
    
    
    # Calculate mean squared error
    mse = mean_squared_error(y_test, y_pred)  # Now this will work
    mse_list.append(mse)

# Print the list of mean squared errors
print(mse_list)

#  Report the mean and the standard deviation of the mean squared errors.

mean_mse = np.mean(mse_list)  # Calculate the mean of the MSEs
std_mse = np.std(mse_list)    # Calculate the standard deviation of the MSEs

# Print the results
print(f"Mean of Mean Squared Errors: {mean_mse}")
print(f"Standard Deviation of Mean Squared Errors: {std_mse}")

[80.23801766930869, 129.9704687660414, 136.32393083113305, 127.79804742556703, 116.17345298002336, 138.3066689915849, 156.4241336811637, 120.67092327657848, 145.97599292635377, 127.77449989762476, 122.07913399867223, 124.19341688530976, 130.7506097950394, 114.62580537165393, 108.67745701877598, 106.45151665272991, 110.66608831523355, 107.93373095358545, 70.7099558851428, 103.77434207260274, 129.6713420948738, 111.53989919719618, 125.35096221957988, 133.80412534519596, 125.56184456486774, 112.27613909327106, 145.45316557540502, 132.84602855040126, 122.77309374981878, 123.51489216128934, 141.07156630615867, 86.32667997133046, 111.31850894957364, 125.20597952633936, 138.50479752160004, 136.0215086624037, 142.9267604900342, 114.73920858149144, 128.79633943577815, 133.35154976500274, 126.04485970807585, 147.9451673438497, 141.05390877929082, 134.1729503296255, 141.8811741844559, 132.38325132881317, 138.3502398132462, 127.69281397614108, 128.02964451642623, 139.13596229772332]
Mean of Mean S

#### the mean has been signgactly reduced to 125.145 after adding the new hidden layers